Changes to FrequencyMapAgent to solve the JSON bug

Krebrov001 opened the pull request:
The FrequencyMapAgent adds some information to the host report about the frequency map that cannot be parsed as YAML.

This fixes the frequency map so that it becomes valid JSON, which can then be parsed by the python interpreter.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

- Relates to #1483 
- Fixes #1483

Krebrov001 committed:
Changes to FrequencyMapAgent to solve the JSON bug

WIP

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Added new tests to FrequencyMapAgentTest.cpp

- FrequencyMapAgentTest.test_hash_freq_map
- FrequencyMapAgentTest.test_default_freq_hash
- FrequencyMapAgentTest.test_both_map_and_set
- FrequencyMapAgentTest.test_neither_map_nor_set

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on src/FrequencyMapAgent.hpp:
```diff
@@ -79,6 +79,11 @@ namespace geopm
             static std::unique_ptr<Agent> make_plugin(void);
             static std::vector<std::string> policy_names(void);
             static std::vector<std::string> sample_names(void);
+            void set_maps(const std::map<uint64_t, double>& hash_freq_map,
+                          const std::set<uint64_t>& default_freq_hash) {
+                m_hash_freq_map = hash_freq_map;
+                m_default_freq_hash = default_freq_hash;
+            }
```
I think it would be better to test this through our integration tests than to introduce a new class method that is only used in testing.

Typically when we're testing the behavior of agents, this is done through an integration test.

This would also illustrate the indentation issues I commented on.  You won't see that in a unit test.

bgeltz replied:
Don't do this, do it with a test constructor instead.

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -364,22 +364,33 @@ namespace geopm
     {
         std::vector<std::pair<std::string, std::string> > result;
-        std::ostringstream oss;
-        oss << std::setprecision(M_PRECISION) << std::scientific;
```
This line that uses the ```set_precision``` call is important and it must be preserved.

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -364,22 +364,33 @@ namespace geopm
+        std::map<std::string, Json> temp_map;
+        for (const auto &region : full_map)
+        {
+            // Json object(int(region.first));
+            // std::string key_string = object.dump();
+            // key_string.erase(std::remove(key_string.begin(), key_string.end(), '"'), key_string.end());
+            // temp_map[key_string] = region.second;
```
Remove unused code prior to review.  It distracts from what you are really doing with the changed code.

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -364,22 +364,33 @@ namespace geopm
-            oss << ": " << m_default_freq;
+
+        std::map<uint64_t, double> full_map(m_hash_freq_map);
+        std::for_each(m_default_freq_hash.begin(), m_default_freq_hash.end(), [&](uint64_t key)
+        {
+            full_map.insert({ key, m_default_freq });
+        });
```
I'd avoid `std::for_each`.  This is much simpler:

```suggestion      
        for (const auto &region : m_default_freq_hash) {
            full_map.insert({region, m_default_freq});
        }
```

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -364,22 +364,33 @@ namespace geopm
+            // key_string.erase(std::remove(key_string.begin(), key_string.end(), '"'), key_string.end());
+            // temp_map[key_string] = region.second;
+            std::ostringstream key_ss;
+            key_ss << "0x" << std::hex << std::setfill('0') << std::setw(16) << std::fixed;
```
```suggestion
            key_ss << "\n    0x" << std::hex << std::setfill('0') << std::setw(16) << std::fixed;
```

This needs to remain so that the indentation in the report is still correct.

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -364,22 +364,33 @@ namespace geopm
         }
-        oss << "\n";
-        result.push_back(std::make_pair("Frequency map", oss.str()));
+        Json my_json = Json(temp_map);
```
I'd prefer to see the original method of using a single `std::ostringstream` used here instead of splitting up the region and the frequency strings into separate streams.

You should be able to avoid that `erase` call, and it's overall less code.  The way you have it here will also result in improper indentation in the report which will cause a parsing problem in the `RawReport` Python code.

The resulting output string from the `std::ostringstream` could still be parsed by json11 if that is desired.

bgeltz commented on test/FrequencyMapAgentTest.cpp:
```diff
@@ -485,3 +486,112 @@ TEST_F(FrequencyMapAgentTest, validate_policy)
+    for (auto& key_value : result) {
+        key_value.second.erase(std::remove(key_value.second.begin(), key_value.second.end(), '"'), key_value.second.end());
+        // std::cerr << key_value.first << " : " << key_value.second << std::endl;
+        // std::cerr << key_value.first << " : " << reference_map << std::endl;
```
```suggestion
```

Krebrov001 committed:
Change the FrequencyMap agent to take the maps into test constructors

Remove the `set_maps()` method, and transfer all of it's functionality
into the test constructors.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Add new method `replace_all_substrings()` to `Helper.hpp`

And add new tests for it.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

Krebrov001 committed:
Implement review feedbacks

- In the frequency map, the key is now a string, and the value is now a
  `Json` object which has been constructed from a double value.
- Modify the relevant tests in FrequencyMapAgentTest to get away from
  scientific notation, and instead format the numbers just as they are
  printed to stdout.
- Replace all occurences of `"nan"` with `"null"` in the final result string.
- Get rid of `const int M_PRECISION;`
- Remove extra line `key_string.erase()`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on service/src/geopm/Helper.hpp:
```diff
@@ -89,6 +89,16 @@ namespace geopm
+    /// @param [in] old_substring The substring to be replaced.
+    /// @param [in] new_substring The new substring to replace the old one with.
+    /// @return bool True if one or more replacements have been made.
+    bool replace_all_substrings(std::string &source_dest,
+                                const std::string &old_substring,
+                                const std::string &new_substring);
```
The function and its implementation should be removed.  It is not used anywhere in the implementation for the bug fix.

Krebrov001 replied:
It would be useful later though. I spend several hours on this implementation.

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -364,22 +372,26 @@ namespace geopm
+            key_ss << "0x" << std::hex << std::setfill('0') << std::setw(16) << std::fixed;
+            key_ss << region.first;
+            std::string key_string = key_ss.str();
+            key_string.erase(std::remove(key_string.begin(), key_string.end(), '"'), key_string.end());
```
Is this line still needed?  We want the keys (which are region hashes in this case) to be treated as strings.  Please add a comment here on why this line is still needed if it is.

Krebrov001 replied:
You're right, this line is extraneous, I'll remove it.

bgeltz commented on test/FrequencyMapAgentTest.cpp:
```diff
@@ -485,3 +486,108 @@ TEST_F(FrequencyMapAgentTest, validate_policy)
+
+    auto result = frequency_agent.report_host();
+    for (auto& key_value : result) {
+        key_value.second.erase(std::remove(key_value.second.begin(), key_value.second.end(), '"'), key_value.second.end());
```
Same question here as before: what purpose does this line serve?  The map is empty.

Krebrov001 replied:
This line removes the extra `"` characters from the string. Previously the keys used to be strings surrounded with quotes: `"0x01234567"` Now they are just `0x01234567`

Krebrov001 committed:
Remove `replace_all_substrings()`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on src/FrequencyMapAgent.cpp:
```diff
@@ -364,22 +372,25 @@ namespace geopm
         }
-        oss << "\n";
-        result.push_back(std::make_pair("Frequency map", oss.str()));
+        Json my_json = Json(temp_map);
+
+        std::string frequency_map_data = my_json.dump();
+        result.push_back(std::make_pair("Frequency map", frequency_map_data));
```
```suggestion
        result.push_back(std::make_pair("Frequency map", Json(temp_map).dump()));
```

bgeltz commented on test/FrequencyMapAgentTest.cpp:
```diff
@@ -485,3 +486,108 @@ TEST_F(FrequencyMapAgentTest, validate_policy)
+
+    auto result = frequency_agent.report_host();
+    for (auto& key_value : result) {
+        key_value.second.erase(std::remove(key_value.second.begin(), key_value.second.end(), '"'), key_value.second.end());
```
```suggestion
```

bgeltz commented on test/FrequencyMapAgentTest.cpp:
```diff
@@ -332,12 +339,6 @@ TEST_F(FrequencyMapAgentTest, split_policy)
 #endif
 }
 
-TEST_F(FrequencyMapAgentTest, name)
-{
-    EXPECT_EQ("frequency_map", m_agent->plugin_name());
-    EXPECT_NE("bad_string", m_agent->plugin_name());
-}
-
```
Why did this move up?  Please put it back.  It is not related to the other changes you are making.

bgeltz commented on test/Makefile.mk:
```diff
@@ -212,13 +212,17 @@ GTEST_TESTS = test/gtest_links/AccumulatorTest.empty \
               test/gtest_links/FrequencyGovernorTest.validate_policy \
+              test/gtest_links/FrequencyMapAgentTest.name \
               test/gtest_links/FrequencyMapAgentTest.adjust_platform_map \
               test/gtest_links/FrequencyMapAgentTest.adjust_platform_uncore \
-              test/gtest_links/FrequencyMapAgentTest.name \
+              test/gtest_links/FrequencyMapAgentTest.split_policy \
               test/gtest_links/FrequencyMapAgentTest.enforce_policy \
               test/gtest_links/FrequencyMapAgentTest.policy_to_json \
-              test/gtest_links/FrequencyMapAgentTest.split_policy \
```
Please put these (the `name` and `split_policy` tests) back in the order they were previously in.  We try to keep things arranged alphabetically by class.

bgeltz commented on test/Makefile.mk:
```diff
@@ -192,6 +192,10 @@ GTEST_TESTS = test/gtest_links/AccumulatorTest.empty \
               test/gtest_links/FrequencyMapAgentTest.policy_to_json \
               test/gtest_links/FrequencyMapAgentTest.split_policy \
               test/gtest_links/FrequencyMapAgentTest.validate_policy \
+              test/gtest_links/FrequencyMapAgentTest.test_hash_freq_map \
+              test/gtest_links/FrequencyMapAgentTest.test_default_freq_hash \
+              test/gtest_links/FrequencyMapAgentTest.test_both_map_and_set \
+              test/gtest_links/FrequencyMapAgentTest.test_neither_map_nor_set \
```
```suggestion
              test/gtest_links/FrequencyMapAgentTest.report_hash_freq_map \
              test/gtest_links/FrequencyMapAgentTest.report_default_freq_hash \
              test/gtest_links/FrequencyMapAgentTest.report_both_map_and_set \
              test/gtest_links/FrequencyMapAgentTest.report_neither_map_nor_set \
```

Then alphabetize the list above here.

bgeltz reviewed (changes requested):
Please take a look at my comment.

Also, were you going to fix the PR description?  It still says WIP.

Krebrov001 committed:
Renaming the new added tests in FrequencyMapAgenTest

- test_hash_freq_map to report_hash_freq_map
- test_default_freq_hash to report_default_freq_hash
- test_both_map_and_set to report_both_map_and_set
- test_neither_map_nor_set to report_neither_map_nor_set

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

bgeltz commented on test/Makefile.mk:
```diff
@@ -190,6 +190,10 @@ GTEST_TESTS = test/gtest_links/AccumulatorTest.empty \
               test/gtest_links/FrequencyMapAgentTest.name \
               test/gtest_links/FrequencyMapAgentTest.enforce_policy \
               test/gtest_links/FrequencyMapAgentTest.policy_to_json \
+              test/gtest_links/FrequencyMapAgentTest.report_hash_freq_map \
+              test/gtest_links/FrequencyMapAgentTest.report_default_freq_hash \
+              test/gtest_links/FrequencyMapAgentTest.report_both_map_and_set \
+              test/gtest_links/FrequencyMapAgentTest.report_neither_map_nor_set \
```
```suggestion
              test/gtest_links/FrequencyMapAgentTest.report_both_map_and_set \
              test/gtest_links/FrequencyMapAgentTest.report_default_freq_hash \
              test/gtest_links/FrequencyMapAgentTest.report_hash_freq_map \
              test/gtest_links/FrequencyMapAgentTest.report_neither_map_nor_set \
```

bgeltz replied:
That's the proper alphabetical order.
